[tzqgav10] iP

tzqgav10 opened the pull request.

tzqgav10 committed:
rename chatbot and add greeting and exit messages

tzqgav10 committed:
Add scanner to take in user input

User can give input to Chatbot and it will respond according to the
input. It can either add task, print the list of tasks or exit when
"bye" is entered as input.

tzqgav10 committed:
Add Task class to handle list of tasks

The Task class allows us to add tasks to a list. There is a method that
prints the tasks in order.

tzqgav10 committed:
Change if-else statements to switch statements, add new cases

Using switch statements look much easier for us to understand and
handle each of the cases depending on user input. Two new cases, mark
and unmark would have an index behind the input which is why an if
statement is used to check for the specific input first.

tzqgav10 committed:
Add task completion status and method to change status

The two new methods are able to change the status of the tasks if
needed.

RaidaNUS commented on src/main/java/Peira.java:
```diff
@@ -0,0 +1,53 @@
+                System.out.println("    added: " + line);
+                Task.addTask(line);
+                System.out.println("    ____________________________________________________________");
+                break;
```
Great way to avoid excessive nested if statements!

tzqgav10 committed:
Add 3 new classes which are inherited from the main task class

The new classes allow the chatbot to create todo, deadlines and events
while being able to use the methods of the main task class.

Isaaclks7 commented on src/main/java/Task.java:
```diff
@@ -23,10 +23,7 @@ public static void printTasks() {
         int index = 1;
         System.out.println("    Here are the tasks in your lists:");
         for (int i = 0; i < totalTasks; i++) {
```
Can consider renaming i to taskIndex or similar for readability

Isaaclks7 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,24 @@
+public class Deadline extends Task {
+
+    protected String by;
+
+    public Deadline(String description, String by) {
+        super(description);
+        this.by = by;
+    }
+
+    public static void addTask(Task task) {
+        entireList[totalTasks] = task;
+        totalTasks++;
+        System.out.println("    ____________________________________________________________");
+        System.out.println("    Got it. I've added this task:");
+        System.out.println("      " + task);
+        System.out.println("    Now you have " + Task.totalTasks + " tasks in the list.");
+        System.out.println("    ____________________________________________________________");
+    }
```
Correct naming conventions for your class, methods and constructor.

Isaaclks7 commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,27 @@
+        this.to = to;
+    }
+
+    public static void addTask(Task task) {
+        entireList[totalTasks] = task;
+        totalTasks++;
+        System.out.println("    ____________________________________________________________");
+        System.out.println("    Got it. I've added this task:");
+        System.out.println("      " + task);
+        System.out.println("    Now you have " + Task.totalTasks + " tasks in the list.");
+        System.out.println("    ____________________________________________________________");
+    }
```
Should place this in Task.java instead if it is the same as in your other subclasses, Todo and Deadline.

tzqgav10 replied:
Thanks for mentioning this!

Isaaclks7 commented on src/main/java/Todo.java:
```diff
@@ -0,0 +1,23 @@
+public class Todo extends Task {
+    protected boolean isDone;
```
Can remove this as it already exists in your Task class.

Isaaclks7 reviewed (commented):
Good organisation of your different types of classes. Can enhance by placing common methods or variables in your parent class.

RaidaNUS commented on src/main/java/Peira.java:
```diff
@@ -0,0 +1,72 @@
+                String to = line.substring(toIndex + 5);
+                Event.addTask(new Event(description, by, to));
+            }
+            switch (line) {
+            case "bye":
+                System.out.println("    ____________________________________________________________");
+                System.out.println("    Bye. Hope to see you again soon!");
+                System.out.println("    ____________________________________________________________");
+                return; // to exit the program entirely
+            case "list":
+                System.out.println("    ____________________________________________________________");
+                Task.printTasks();
+                System.out.println("    ____________________________________________________________");
+                break;
+            case "mark":
+                System.out.println("    ____________________________________________________________");
+                Task.markDone(taskIndex);
+                System.out.println("    ____________________________________________________________");
+                break;
+            case "unmark":
+                System.out.println("    ____________________________________________________________");
+                Task.markUndone(taskIndex);
+                System.out.println("    ____________________________________________________________");
+                break;
+            default:
+                // System.out.println("    Instructions unclear. Please input a valid command!");
+                break;
+            }
+        }
+    }
+}
```
Great way to substitute usage of if-else statement

RaidaNUS reviewed (commented):
Overall Great!

tzqgav10 committed:
Refactor code by extraction and make code more readable

Extracted code to handle printLine(), todo, event, deadline
for better readability.

tzqgav10 committed:
Add automated text UI testing

Allows us to check for any difference in expected and output

tzqgav10 committed:
Add exceptions to handle exceptional cases for commands entered by user

Users can now see relevant help messages whenever they enter wrong
commands

tzqgav10 committed:
Merge branch 'branch-Level-5'

tzqgav10 committed:
Divide classes into packages

tzqgav10 committed:
Add delete function

Delete function allows us to remove tasks from the list. There are
proper exceptions thrown when the index entered by user is out of
bounds.

tzqgav10 committed:
Add file storage for tasks with OS-independent paths

Implemented file storage functionality to save tasks automatically
whenever the task list changes. Added support for loading tasks from a
file when the chatbot starts up. Used `java.nio.file.Paths` and
`System.getProperty("user.home")` to ensure OS-independent file paths.
Handled cases where the data file or directory does not exist at
startup. Integrated polymorphism to support saving and loading of
`Todo`, `Deadline`, and `Event` tasks.

tzqgav10 committed:
Merge branch 'branch-Level-7'

tzqgav10 committed:
Add saving of files when tasks are removed

xenthm commented on src/main/java/peira/task/Task.java:
```diff
@@ -0,0 +1,139 @@
+    // add task method
+    public static void addTask(Task task) {
+        entireList.add(task); // integrates collection to add task
+        System.out.println("    ____________________________________________________________");
+        System.out.println("    Got it. I've added this task:");
+        System.out.println("      " + task);
+        System.out.println("    Now you have " + Task.entireList.size() + " tasks in the list.");
+        System.out.println("    ____________________________________________________________");
```
I like the use of indentations in your messages to make them more visually structured to users, but perhaps consider using a constant to represent it (i.e. `INDENT`) to avoid [magic numbers](https://nus-cs2113-ay2425s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-magic-numbers). Same goes for the lines.

xenthm commented on src/main/java/peira/task/Deadline.java:
```diff
@@ -0,0 +1,28 @@
+        this.by = by;
+    }
+    // method to initialize description and by for deadline
+    public static Deadline fromFileString(String fileString) {
+        boolean isDone = fileString.charAt(4) == 'X';
+        int indexOfDescription = fileString.indexOf(" (");
+        String description = fileString.substring(6, indexOfDescription);
+        int indexOfBy = fileString.indexOf("by: ");
+        int endIndexOfBy = fileString.indexOf(")");
+        String by = fileString.substring(indexOfBy + 4, endIndexOfBy);
+        Deadline deadline = new Deadline(description, by);
+        deadline.isDone = isDone;
+        return deadline;
+    }
```
Great job naming everything descriptively in this method and others like it!

xenthm commented on src/main/java/peira/task/Task.java:
```diff
@@ -0,0 +1,139 @@
+        }
+        entireList.get(index-1).isDone = true;
+        System.out.println("    Nice! I've marked this task as done:");
+        System.out.println("     " + entireList.get(index-1).toString());
```
Do try to use the formatting for white space outlined in [the Java coding standards for this course](https://se-education.org/guides/conventions/java/index.html#layout). This also applies to some other parts of your codebase, try to take a look!

tzqgav10 committed:
Refactor the code to make it more OOP

Added classes for commands, ui, parser and storage to refactor code,
for enhanced readability. Fixed some bugs that was causing the code
to be unable to read the status of todo tasks.

tzqgav10 committed:
Taught chatbot how to understand date and times

Chatbot can read dates for deadlines and events in a specific format. If
it does not understand, the chatbot will print the original input.

tzqgav10 committed:
Add a search functionality

Chatbot will be able to search tasks associated with the provided
keyword and print out those list of tasks.

tzqgav10 committed:
Add Javadoc header comments to public classes/methods

tzqgav10 committed:
Merge pull request #1 from tzqgav10/branch-Level-8

Merge level 8 with master

tzqgav10 committed:
Merge branch 'master' into branch-Level-9

tzqgav10 committed:
Merge branch 'master' into branch-A-JavaDoc

tzqgav10 committed:
Merge pull request #2 from tzqgav10/branch-Level-9

Merge level 9 with master

tzqgav10 committed:
Merge branch 'master' into branch-A-JavaDoc

tzqgav10 committed:
Merge pull request #3 from tzqgav10/branch-A-JavaDoc

Merge A-JavaDoc to master

tzqgav10 committed:
Add constants to avoid magic numbers

tzqgav10 committed:
Add User Guide for Chatbot

tzqgav10 committed:
Change formatting of User Guide
